src/TestResults/match-by-context.ts

```diff
@@ -176,11 +188,28 @@ interface FallbackMatchResult<C extends ContextType> {
 }
 
 type ClassicMatchType = 'by-name' | 'by-location';
+type MaybeTestResult<C extends boolean> = C extends true ? TestResult : undefined;
 
 const ContextMatch = (): ContextMatchAlgorithm => {
+  const onMatch = <C extends boolean>(
+    tNode: NodeType<ItBlock>,
+    aNode: NodeType<TestAssertionStatus>,
+    event: MatchEvent,
+    generateResult: C
+  ): MaybeTestResult<C> => {
+    tNode.addEvent(event);
+    aNode.addEvent(event);
+    aNode.attrs.range = tNode.attrs.range;
+
+    if (generateResult && isDataNode(tNode) && isDataNode(aNode)) {
+      return toMatchResult(tNode, aNode, event) as MaybeTestResult<C>;
+    }
+    return undefined as MaybeTestResult<C>;
+  };
```

Comment:
Would it maybe make sense to split it up into two separate handler (one calling the other) instead of doing the (still elegant) type workaround?

Reply:
I think you are right, no need to do any clever type-trick, separate methods are just fine and maybe even with less mental overhead.